File: src/test/java/org/tindalos/principle/domain/checker/LayeringTest.java
package org.tindalos.principle.domain.checker;

import static org.junit.Assert.assertEquals;

import java.util.Arrays;
import java.util.List;
import java.util.Set;

import org.junit.BeforeClass;
import org.junit.Test;
import org.tindalos.principle.domain.core.DesignQualityCheckConfiguration;
import org.tindalos.principle.domain.core.ListConverter;
import org.tindalos.principle.domain.coredetector.CheckResult;
import org.tindalos.principle.domain.detector.layering.LayerReference;
import org.tindalos.principle.domain.detector.layering.LayerViolationsResult;
import org.tindalos.principle.domain.expectations.DesignQualityExpectations;
import org.tindalos.principle.domain.expectations.Layering;
import org.tindalos.principle.infrastructure.di.PoorMansDIContainer;
import org.tindalos.principle.infrastructure.plugin.Checks;

import com.google.common.collect.Sets;

public class LayeringTest {

    private DesignQualityCheckConfiguration designQualityCheckConfiguration;
    private DesignQualityCheckService designQualityCheckService;
    private DesignQualityExpectations checks = prepareChecks();

    @BeforeClass
    public static void setup() {
        TestFixture.setLogger();
    }

    private void init(String basePackage) {
        designQualityCheckService = PoorMansDIContainer.getDesignCheckService(basePackage);
        designQualityCheckConfiguration = new DesignQualityCheckConfiguration(checks, basePackage);
    }

    @Test
    public void simple() {
        
        List<LayerReference> result = run("org.tindalos.principletest.layering.simple");
        
        Set<LayerReference> expected = Sets.newHashSet(
                new LayerReference("org.tindalos.principletest.layering.simple.domain", "org.tindalos.principletest.layering.simple.app")
                ,new LayerReference("org.tindalos.principletest.layering.simple.domain", "org.tindalos.principletest.layering.simple.infrastructure")
                ,new LayerReference("org.tindalos.principletest.layering.simple.app", "org.tindalos.principletest.layering.simple.infrastructure")
                );
        assertEquals(expected, Sets.newHashSet(result));
    }
    
    @Test
    public void deeper() {
        
        List<LayerReference> result = run("org.tindalos.principletest.layering.deeper");
        
        Set<LayerReference> expected = Sets.newHashSet(
                new LayerReference("org.tindalos.principletest.layering.deeper.domain.aaa", "org.tindalos.principletest.layering.deeper.app.bbb.ccc")
                );
        assertEquals(expected, Sets.newHashSet(result));
    }

    
    private List<LayerReference> run(String basePackage) {
        init(basePackage);
        DesignQualityCheckResults checkResults = designQualityCheckService.analyze(designQualityCheckConfiguration);
        List<CheckResult> resultList = checkResults.resultList();
        assertEquals(1, resultList.size());
        LayerViolationsResult result = (LayerViolationsResult) resultList.get(0);
        return ListConverter.convert(result.violations());
    }

    private static DesignQualityExpectations prepareChecks() {
        Checks checks = new Checks();
        checks.layering_$eq(layering());
        return checks;
    }

    private static Layering layering() {
        Layering layering = new Layering();
        layering.setLayers(Arrays.asList("infrastructure","app","domain"));
        return layering;
    }


}

File: src/test/scala/org/tindalos/principle/domain/checker/LayeringTest.scala
package org.tindalos.principle.domain.checker

import org.junit.Assert.assertEquals
import org.junit._
import org.tindalos.principle.domain.AnalysisInput
import org.tindalos.principle.domain.core.AnalysisPlan
import org.tindalos.principle.domain.analyzers.layering.{LayerReference, LayerViolationsResult}
import org.tindalos.principle.domain.constraints._
import org.tindalos.principle.domain.core.packages.PackageWithMetrics
import org.tindalos.principle.infrastructure.JDependBasedPackageListBuilder
import org.tindalos.principle.infrastructure.di.PoorMansDIContainer

import scala.collection.JavaConverters._

class LayeringTest {

  var plan: AnalysisPlan = null
  var analysisRunner = PoorMansDIContainer.buildAnalysisRunner()
  var expectations: Constraints = prepareChecks()

  @Before
  def setup() = {
    TestFixture.setLogger()
  }

  @Test
  def simple() = {

    val result = run("org.tindalos.principletest.layering.simple")
    val expected = Set(new LayerReference("org.tindalos.principletest.layering.simple.domain", "org.tindalos.principletest.layering.simple.app"),
      new LayerReference("org.tindalos.principletest.layering.simple.domain", "org.tindalos.principletest.layering.simple.infrastructure"),
      new LayerReference("org.tindalos.principletest.layering.simple.app", "org.tindalos.principletest.layering.simple.infrastructure"))
    assertEquals(expected, result.asScala.toSet)
  }

  @Test
  def deeper() = {

    val result = run("org.tindalos.principletest.layering.deeper")
    val expected = Set(new LayerReference("org.tindalos.principletest.layering.deeper.domain.aaa", "org.tindalos.principletest.layering.deeper.app.bbb.ccc"))
    assertEquals(expected, result.asScala.toSet)
  }

  def init(basePackage: String) = {
    plan = new AnalysisPlan(expectations, basePackage)
  }

  private def run(basePackage: String) = {
    init(basePackage)
    val packageListProducer = new JDependBasedPackageListBuilder(basePackage)
    val packageList = packageListProducer.build()
    val result = analysisRunner.run(new AnalysisInput(packageList.map(p => p: PackageWithMetrics).asJava, Set.empty.asJava, plan))
    assertEquals(1, result.size())
    result.get(0).asInstanceOf[LayerViolationsResult].violations
  }

  private def prepareChecks() = Constraints.builder().layering(layering()).build()

  private def layering() = new Layering(java.util.List.of("infrastructure", "app", "domain"), 0)

}